### PR TITLE
secure storage: update head message when info->dataSize update

### PR DESCRIPTION
**Pull request by lackan**
When one invokes TEE_WriteObjectData and write some data into the secure storage file, the data size of the file may change, but currently it will not update the head message in persistent objects. This commit will fix this problem by updating head message when info->dataSize update.

Relates to:
#814 
Output:(test on FVP)
```
Object data size : 50 and position : 50
Object data size : 50 and position : 0
```

Test for xtest have add to OP-TEE/optee_test#151

Signed-off-by: Guanchao Liang <liang.guanchao@linaro.org>
Reviewed-by: Joakim Bech <joakim.bech@linaro.org>
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

File: core/tee/tee_svc_storage.c [hunk 1]
```diff
@@ -297,6 +297,54 @@ static TEE_Result tee_svc_storage_read_head(struct tee_ta_session *sess,
...
+	res = fops->seek(o->fh, 0, TEE_DATA_SEEK_SET, NULL);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	/* read head */
+	bytes = sizeof(struct tee_svc_storage_head);
+	res = fops->read(o->fh, &head, &bytes);
+	if (res != TEE_SUCCESS) {
+		if (res == TEE_ERROR_CORRUPT_OBJECT)
+			EMSG("Head corrupt.\n");
```
Comment (jbech-linaro) [inline core/tee/tee_svc_storage.c]: `\n` not needed.

**Comment by jbech-linaro** [comment]
With or without my comment fixed.
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: core/tee/tee_svc_storage.c [hunk 2]
```diff
@@ -297,6 +297,54 @@ static TEE_Result tee_svc_storage_read_head(struct tee_ta_session *sess,
...
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = fops->seek(o->fh, 0, TEE_DATA_SEEK_SET, NULL);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	/* read head */
+	bytes = sizeof(struct tee_svc_storage_head);
+	res = fops->read(o->fh, &head, &bytes);
```
Comment (jenswi-linaro) [inline core/tee/tee_svc_storage.c]: You could skip the read step if you'd only update the position field in the file.

File: core/tee/tee_svc_storage.c (hunk 1 above)
Comment (lackan) [inline core/tee/tee_svc_storage.c]: OK

File: core/tee/tee_svc_storage.c (hunk 2 above)
Comment (jenswi-linaro) [inline core/tee/tee_svc_storage.c]: I think we're good without that assert.

Comment (lackan) [inline core/tee/tee_svc_storage.c]: OK

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/tee/tee_svc_storage.c [hunk 3]
```diff
@@ -1069,8 +1098,10 @@ TEE_Result syscall_storage_obj_write(unsigned long obj, void *data, size_t len)
 		goto exit;
 
 	o->info.dataPosition += len;
-	if (o->info.dataPosition > o->info.dataSize)
+	if (o->info.dataPosition > o->info.dataSize) {
 		o->info.dataSize = o->info.dataPosition;
```
Comment (etienne-lms) [inline core/tee/tee_svc_storage.c]: update `dataPosition` and `dataSize` only if successfully updated the head ?

Comment (jenswi-linaro) [inline core/tee/tee_svc_storage.c]: Good catch!

File: core/tee/tee_svc_storage.c [hunk 4]
```diff
@@ -1069,8 +1097,12 @@ TEE_Result syscall_storage_obj_write(unsigned long obj, void *data, size_t len)
 		goto exit;
 
 	o->info.dataPosition += len;
```
Comment (jenswi-linaro) [inline core/tee/tee_svc_storage.c]: I don't think this should be updated if we can't update head below.

Comment (lackan) [inline core/tee/tee_svc_storage.c]: But the write operation have overwrote the data, the data position actually have changed? So we ignore this overwrite?

Comment (jenswi-linaro) [inline core/tee/tee_svc_storage.c]: Hmm, this is actually a problem with the way we're layering this. The update in the lower layer is already committed and what we would need now is unrolling it. This problem is out of scope of this P-R so please ignore my previous comment for now.

Comment (lackan) [inline core/tee/tee_svc_storage.c]: OK, so I just keep it as now.

**Comment by jforissier** [comment]
@lackan did @jenswi-linaro give is R-b? I can't see it in the discussion, yet you added it to your first commit. @jenswi-linaro please let us know if you're OK with the current state.

Also, @lackan could you please add a test to xtest for this case?

**Comment by jenswi-linaro** [comment]
Fine with me:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by lackan** [comment]
@jforissier Yes, I will add a test to xtest soon.

**Comment by jforissier** [comment]
Rebased on top of master and merged. Thanks @lackan.
